From 13bae86b4d3ed01f885e111eda4224a2985ed9a4 Mon Sep 17 00:00:00 2001
From: Md Sadre Alam <mdalam@codeaurora.org>
Date: Tue, 21 Jul 2020 21:19:58 +0530
Subject: [PATCH 4/7] mtd: rawnand: qcom: Read QPIC version

This change will add support to read QPIC version.
QPIC version V2.0 onwards some new register introduced
in QPIC. So based on hw_version we will update those
register.

Signed-off-by: Md Sadre Alam <mdalam@codeaurora.org>
Change-Id: Iadc95a12145e03b8edf0deb6537b819ad4baec22

Pick from https://git.codelinaro.org/clo/qsdk/oss/kernel/linux-ipq-5.4/-/commit/bfd90c9e9e90c0fa9150d25d2b2e6d4d321610f0

Signed-off-by: hzy <hzyitc@outlook.com>
---
 drivers/mtd/nand/raw/qcom_nandc.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/mtd/nand/raw/qcom_nandc.c b/drivers/mtd/nand/raw/qcom_nandc.c
index 361ee116422f..e19074062348 100644
--- a/drivers/mtd/nand/raw/qcom_nandc.c
+++ b/drivers/mtd/nand/raw/qcom_nandc.c
@@ -442,6 +442,7 @@ struct qcom_nand_controller {
 
 	u32 cmd1, vld;
 	const struct qcom_nandc_props *props;
+	u32 hw_version;
 };
 
 /*
@@ -2539,6 +2540,7 @@ static int qcom_nand_attach_chip(struct nand_chip *chip)
 	int ecc_mode = 1;
 	int num_addr_cycle = 5, dsbl_sts_aftr_write = 0;
 	int wr_rd_bsy_gap = 2, recovery_cycle = 7;
+	u32 version_reg;
 
 	/* controller only supports 512 bytes data steps */
 	ecc->size = NANDC_STEP_SIZE;
@@ -2552,6 +2554,16 @@ static int qcom_nand_attach_chip(struct nand_chip *chip)
 	if (chip->base.eccreq.strength >= 8)
 		ecc->strength = 8;
 
+	/* Read QPIC version register */
+	version_reg = (NAND_VERSION + 0x4000);
+	nandc->hw_version = nandc_read(nandc, version_reg);
+	pr_info("QPIC controller hw version Major:%d, Minor:%d\n",
+			((nandc->hw_version & NAND_VERSION_MAJOR_MASK)
+			 >> NAND_VERSION_MAJOR_SHIFT),
+			((nandc->hw_version & NAND_VERSION_MINOR_MASK)
+			 >> NAND_VERSION_MINOR_SHIFT));
+	nandc->hw_version = ((nandc->hw_version & NAND_VERSION_MAJOR_MASK)
+			>> NAND_VERSION_MAJOR_SHIFT);
 	/*
 	 * Each CW has 4 available OOB bytes which will be protected with ECC
 	 * so remaining bytes can be used for ECC.
-- 
2.25.1

